From 5519446777e10751ea2ed0cbd05402b72043ad1a Mon Sep 17 00:00:00 2001
From: Pratham Patel <ppatel@ciq.com>
Date: Mon, 11 Aug 2025 21:44:44 +0530
Subject: [PATCH] cifs: potential buffer overflow in handling symlinks

jira VULN-49620
cve CVE-2022-49058
commit-author Harshit Mogalapalli <harshit.m.mogalapalli@oracle.com>
commit 64c4a37ac04eeb43c42d272f6e6c8c12bfcf4304

Smatch printed a warning:
	arch/x86/crypto/poly1305_glue.c:198 poly1305_update_arch() error:
	__memcpy() 'dctx->buf' too small (16 vs u32max)

It's caused because Smatch marks 'link_len' as untrusted since it comes
from sscanf(). Add a check to ensure that 'link_len' is not larger than
the size of the 'link_str' buffer.

Fixes: c69c1b6eaea1 ("cifs: implement CIFSParseMFSymlink()")
	Signed-off-by: Harshit Mogalapalli <harshit.m.mogalapalli@oracle.com>
	Reviewed-by: Ronnie Sahlberg <lsahlber@redhat.com>
	Signed-off-by: Steve French <stfrench@microsoft.com>
(cherry picked from commit 64c4a37ac04eeb43c42d272f6e6c8c12bfcf4304)
	Signed-off-by: Pratham Patel <ppatel@ciq.com>
---
 fs/cifs/link.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/fs/cifs/link.c b/fs/cifs/link.c
index 94dab4309fbb..85d30fef98a2 100644
--- a/fs/cifs/link.c
+++ b/fs/cifs/link.c
@@ -97,6 +97,9 @@ parse_mf_symlink(const u8 *buf, unsigned int buf_len, unsigned int *_link_len,
 	if (rc != 1)
 		return -EINVAL;
 
+	if (link_len > CIFS_MF_SYMLINK_LINK_MAXLEN)
+		return -EINVAL;
+
 	rc = symlink_hash(link_len, link_str, md5_hash);
 	if (rc) {
 		cifs_dbg(FYI, "%s: MD5 hash failure: %d\n", __func__, rc);
-- 
2.47.3

